File: 3.LongestSubstringWithoutRepeatingCharacters.js
var lengthOfLongestSubstring = function (s) {
  const set = new Set();
  let i = 0,
    j = 0,
    result = 0;
  len = s.length;

  while (i < len && j < len) {
    if (!set.has(s[j])) {
      set.add(s[j++]);
      result = Math.max(result, j - i);
    } else {
      set.delete(s[i++]);
    }
  }
  return result;
};

File: 3.LongestSubstringWithoutRepeatingCharacters.java
class Solution {
	public int lengthOfLongestSubstring(String s) {
		if (s == null || s.length() == 0)
			return 0;
		HashMap<Character, Integer> map = new HashMap<>();
		int result = 0;
		for (int i = 0, j = 0; i < s.length(); i++) {
			if (map.containsKey(s.charAt(i))) {
				j = Math.max(j, map.get(s.charAt(i)) + 1);
			}
			map.put(s.charAt(i), i);
			result = Math.max(result, i - j + 1);
		}
		return result;
	}
}